src/posix/platform/spinel.cpp

```diff
@@ -0,0 +1,221 @@
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "platform-posix.h"
+
+#include "posix/platform/spinel.hpp"
+
+#include "common/code_utils.hpp"
+#include "common/new.hpp"
+#include "lib/spinel/spinel_driver.hpp"
+#include "posix/platform/hdlc_interface.hpp"
+#include "posix/platform/radio_url.hpp"
+#include "posix/platform/spi_interface.hpp"
+#include "posix/platform/vendor_interface.hpp"
+
+static ot::Posix::SpinelManager sSpinelManager;
+
+namespace ot {
+namespace Posix {
+
+SpinelManager &GetSpinelManager(void) { return sSpinelManager; }
+
+ot::Spinel::SpinelDriver &GetSpinelDriver(void) { return sSpinelManager.GetSpinelDriver(); }
+
+SpinelManager::SpinelManager(void)
+    : mUrl(nullptr)
+    , mSpinelDriver()
+    , mSpinelInterface(nullptr)
+{
+}
+
+CoprocessorMode SpinelManager::Init(const char *aUrl)
+{
+    bool         swReset;
+    spinel_iid_t iidList[ot::Spinel::kSpinelHeaderMaxNumIid];
+
+    mUrl.Init(aUrl);
+    VerifyOrDie(mUrl.GetPath() != nullptr, OT_EXIT_INVALID_ARGUMENTS);
+
+    GetIidListFromUrl(iidList);
+
+#if OPENTHREAD_POSIX_VIRTUAL_TIME
+    VirtualTimeInit();
+#endif
+
+    mSpinelInterface = CreateSpinelInterface(mUrl.GetProtocol());
+    VerifyOrDie(mSpinelInterface != nullptr, OT_EXIT_FAILURE);
+
+    swReset = !mUrl.HasParam("no-reset");
+
+    mSpinelDriver.Init(*mSpinelInterface, swReset, iidList, OT_ARRAY_LENGTH(iidList));
+
+    otLogDebgPlat("instance init:%p - iid = %d", (void *)&mSpinelDriver, iidList[0]);
+
+    return CoprocessorMode::RCP;
+}
+
+void SpinelManager::Deinit(void)
```

Comment:
Should we consider the adding the destructor function of the class `SpinelManager ` to call `Deinit ()` to release system resources when the program crashes ?